crates/lib/src/qpu/result_data.rs

```diff
@@ -0,0 +1,85 @@
+//! This modules provides types and functions for initializing and working with
+//! data returned from the QPU
+use enum_as_inner::EnumAsInner;
+use num::complex::Complex64;
+use quil_rs::instruction::MemoryReference;
+use std::collections::HashMap;
+
+use qcs_api_client_grpc::models::controller::{
+    readout_values as controller_readout_values, ReadoutValues as ControllerReadoutValues,
+};
+
+/// A row of readout values from the QPU. Each row contains all the values emitted to a
+/// memory reference across all shots.
+#[derive(Debug, Clone, EnumAsInner, PartialEq)]
+pub enum ReadoutValues {
+    /// Integer readout values
+    Integer(Vec<i64>),
+    /// Real numbered readout values
+    Real(Vec<f64>),
+    /// Complex readout values
+    Complex(Vec<Complex64>),
+}
+
+/// This struct encapsulates data returned from the QPU after executing a job.
+#[allow(clippy::module_name_repetitions)]
+#[derive(Debug, Clone, PartialEq)]
+pub struct QpuResultData {
+    /// Mappings of a memory region (ie. "ro\[0\]") to it's key name in `readout_values` (ie. "q0")
+    pub mappings: HashMap<String, String>,
+    /// Mapping of a readout values identifier (ie. "q0") to a set of [`ReadoutValues`]
+    pub readout_values: HashMap<String, ReadoutValues>,
```

Comment:
I could be convinced otherwise - but I think that these (and the QVM equivalents) should remain private (i.e. `pub(crate)` and be accessed via methods. We're going to want to change these as the API evolves and our ideas of result data evolve, and if they're public - we're stuck.

I think the SDK should treat the data coming back from execution as a black box that can be queried for data in various shapes. That way, even if the shape in which they are stored changes - we can continue to uphold our API until we _choose_ to break it.

Reply:
I've implemented this, but a practical problem that arises is whether or not we want to expose constructors for these types. If we do, we are committing to some data shape for initializing the type. We can choose to omit constructors, but that makes the library hard to work with from the outside. Ie. you can't build your own `QpuResultData` for testing without actually submitting a job and retrieving the results.

Reply:
I've implemented constructors, but if we want to avoid that, then the python unit tests targeting these will have to become integration tests.